### ISSUE TITLE
"StableDiffusionControlNetInpaintPipeline" inpainting in edge areas always does not fully fit the mask

### ISSUE THREAD

**Issue by AdamMayor2018** [huggingface/diffusers#3959]
My goal is to use StableDiffusionControlNetInpaintPipeline to realize automatic dressing of models. I used controlnet with openpose and mask image of dress.But I found in edge areas, inpainting will more of less tamper with my mask image.

An example:

original image:
![dress1](https://github.com/huggingface/diffusers/assets/37369009/510381d9-310a-4fd8-acbf-2bc7adc60985)

inpainting result:
![1688609158464](https://github.com/huggingface/diffusers/assets/37369009/74bb40e3-2efc-4de7-92a8-4889edcd4ce8)

Could I get some suggestions to make the edges of the clothes strictly match the mask.

**Comment by sayakpaul** [comment]
@AdamMayor2018 if you could also provide your test code snippet, that would be helpful.

**Comment by AdamMayor2018** [comment]
@sayakpaul I found a solution to avoid some of the bad results by using another canny controlnet only with mask of my target clothes.But somehow it works fine with human generating, but when it comes to background, I see white pixels around the mask border.

![image](https://github.com/huggingface/diffusers/assets/37369009/6f8b70f3-a0fc-46a2-a691-b770fcfdde08)

Besides, I found that lower parameter 'strenght' could reduce this effect, but when inpainting background, parameter 'strength' must be 1.0 for fine results.

```
    def swap_background(self, image, prompt, num_images_per_prompt, num_inference_steps=50, strength=1,guidance_scale=7.5, reverse=True):
        image = image.convert('RGB')
        width, height = image.size
        mask_image = self.base_predictor.segformer_mask_inference(image=image, part="background", reverse=reverse)
        mask_image = cv2.dilate(mask_image, kernel=np.ones((3,3),np.uint8), iterations=1)
        #mask_image = cv2.erode(mask_image, kernel=np.ones((5,5),np.uint8), iterations=2)
        images = self.base_predictor.inpaint_inference(prompt=prompt, init_image=image,
                                              mask_image=Image.fromarray(mask_image),
                                              num_images_per_prompt=num_images_per_prompt,
                                              num_inference_steps=num_inference_steps,
                                              strength=strength,
                                              guidance_scale=guidance_scale, width=width // 8 * 8, height=height // 8 * 8).images
        images = [img.resize((width, height)) for img in images]

        return mask_image, images
```
Now  I'm  thinking using method like dilation or erosion to solve it.Could I have some suggestions?

**Comment by sayakpaul** [comment]
Thanks for sharing your findings! 

> Now I'm thinking using method like dilation or erosion to solve it.Could I have some suggestions?

I like your earlier workarounds better as they're simpler.

**Comment by AdamMayor2018** [comment]
@sayakpaul I think why human inpainting gets better is the segmentation model tends to predict  smaller borders.But when it comes to inpainting background, white pixels show. So my question is that whether the inpainting model could tolerate one or two wrong pixels of the mask?

if I want good inpaintings, paramter 'strenght' must be set to 1, which make border white pixels.
if I set strength lower than 1, white border pixels vanished, but In could only get bad or simple background.了👼

**Comment by AdamMayor2018** [comment]
@sayakpaul Add some bad examples with strenght lower than 1:
![image](https://github.com/huggingface/diffusers/assets/37369009/cb510d53-f564-4371-85a0-38abb249b9c9)
![image](https://github.com/huggingface/diffusers/assets/37369009/9e6c91bd-f759-4843-8373-4f141a5168bc)

why the background is so narrow?

model:chilloutmix
prompt:(masterpeice),(best quality),highres,HDR, indoors background, (wooden furniture:1.3), (European style decoration style:1.4),white wall paper

Besides, I get good results with stable-diffusion-webui with the same settings and params, so I wonder where is the difference?

![image](https://github.com/huggingface/diffusers/assets/37369009/41ad1c99-26cf-4b12-b0ac-2f1a24ae6959)

**Comment by patrickvonplaten** [comment]
Hey @AdamMayor2018,

Curious to try to understand what might be going on here. 

Can you also send me a picture of a mask?

**Comment by AdamMayor2018** [comment]
@patrickvonplaten Hello patrick,
I met a problem that diffusers leaves white pixels when inpainting background.
If I set param strength to lower than 1, I can't get good background results.
If I set param strength to 1.0, results are better, but white pixels appear.
Besides, stable diffusion webui won't have this white pixel, so im confused. See these white pixels in the picture below:

![image](https://github.com/huggingface/diffusers/assets/37369009/4a86ff7f-f665-4b7e-b4d5-79155391ead1)


btw, inpating settings are as follows:

![image](https://github.com/huggingface/diffusers/assets/37369009/e12a8d20-7cdf-44a1-a438-8765030c2359)

**Comment by patrickvonplaten** [comment]
Hey @AdamMayor2018,

A different I see here is that the masked area seems to be "fill" with the background color before being denoised. You could also try this here by replacing this line: https://github.com/huggingface/diffusers/blob/e27500b72c129b4051415688c4653f9142697cbc/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py#L147 with something where instead of setting the masked area to just 0, you set it to your background color. I think this should well :-)

**Comment by Laidawang** [comment]
@patrickvonplaten when i use masked **content=original,** What corresponding changes should I make in the code of diffusers?

**Comment by AdamMayor2018** [comment]
@patrickvonplaten Hi, patrick
I followed your advice to adjust mask image.Heres my code:
```
masked_image = image * (mask < 0.5) 
mean = torch.sum(masked_image, dim=(2, 3)) / (torch.sum(mask, dim=(2, 3)) + 1e-7) 
masked_image = masked_image + mask * mean[:, :, None, None]
```
but I didn't get better results. I cound see the mask is filled with background color.

I think the problem is that my mask image contains white border pixels.
![after](https://github.com/huggingface/diffusers/assets/37369009/0579d7f6-3f55-48e0-b363-53c9ec6bffcd)

And these pixels are not inpainted.
Could I get some suggestions?

**Comment by AdamMayor2018** [comment]
Perhaps stable-diffusion-webui don't have this problem is due to this param?
![image](https://github.com/huggingface/diffusers/assets/37369009/935b88f2-b0fe-4f1a-9444-a0676632d266)

**Comment by AdamMayor2018** [comment]
I want to add some results I tried:
1. Use better segmentation tool.Compared with segformer, Sam produces better mask result.I could get rid of the white pixel border.like these:
![1adc673d5c17e9fd439fe513ddaab8e](https://github.com/huggingface/diffusers/assets/37369009/f2781013-258c-41f2-9853-b993a6f9cc03)
![24fb0b717e9672c3052ed4cdc6ce8e6](https://github.com/huggingface/diffusers/assets/37369009/aa238b62-054e-4726-ad64-6a03be351086)
However there are two things im confused about:
A.is the mask area unchanged completely during the process? I found that person's hair remain unchanged.
![b5333edcfcc7040ef5f972967b2beb1](https://github.com/huggingface/diffusers/assets/37369009/286f7eab-9fd7-4188-bb6d-e7920f0ba403)
see the white pixels under the persons hair.

B.why the strength parameter must be set to 1 if I want good inpainting results?(White pixels will also show up), somehow this never shows up with stable-diffusion-webui.
Compare the two results below:

using diffusers with strength = 1:
![result_0](https://github.com/huggingface/diffusers/assets/37369009/eb980eef-85f8-4354-a198-a81cb67130f1)


using stable-diffusion-webui:
![image](https://github.com/huggingface/diffusers/assets/37369009/3730b3bd-82d0-4d58-9a54-a7ec1ec2b3cb)

why is this?

**Comment by patrickvonplaten** [comment]
Can you also try passing `strength=0.95` or something similarly high - this should fix some issues with the mask border: https://huggingface.co/docs/diffusers/v0.19.3/en/api/pipelines/stable_diffusion/inpaint#diffusers.StableDiffusionInpaintPipeline.__call__.strength

**Comment by AdamMayor2018** [comment]
@patrickvonplaten I already tried strength=0.95, the border seems to get better.But If I wanna draw certain indoors background, strengh=0.95 won't get fine results.
Exmaple:
prompt: "indoors, (European style decoration style:1.2), Ivory long table, light green wallpaper,ceiling lights"
![image](https://github.com/huggingface/diffusers/assets/37369009/52889acb-c88c-4eee-a97f-a6c41e49761b)

The only way to get good indoors background, is to set strength=1, which border problem occurs.
Besides, outdoors background seems fine, like this:
![image](https://github.com/huggingface/diffusers/assets/37369009/b981cef5-c90e-4c5a-8bdd-234d44dc11ff)

**Comment by AdamMayor2018** [comment]
After many tests, I think this is a bug that Controlnet Inpaint pipe could not completely deal with background generation. @patrickvonplaten

**Comment by patrickvonplaten** [comment]
Can you pinpoint the bug to something specific / different?

**Comment by CJEnright** [comment]
I'm also running into this. I'm not positive but I think it has to do with resizing the image and mask before using them as latents, which causes unmasked pixels to get fed into the generation process when it would otherwise be masked out at full size.

This is especially problematic when trying to generate a background for an image that doesn't have a background (for example, the background has previously been removed and made transparent.)

Here's an example using SDXL (although it also happens with 1.5). 

```python
import PIL
import requests
import torch
from io import BytesIO

from diffusers import StableDiffusionXLInpaintPipeline


def download_image(url):
    response = requests.get(url)
    return PIL.Image.open(BytesIO(response.content)).convert("RGB")


img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = download_image(img_url).resize((512, 512))
mask_image = download_image(mask_url).resize((512, 512))

# Invert mask – we want to paint the background given the foreground
inverted_mask = PIL.ImageOps.invert(mask_image)
inverted_mask.save("./inverted_mask.png")

# Mask init_image to only show the dog, in the real world this could be any image without a background that we want to add a background to
white = PIL.Image.new("RGB", init_image.size, (255, 255, 255))
masked_init_image = PIL.Image.composite(
    init_image,
    white,
    PIL.ImageOps.invert(inverted_mask.convert("L"))
)
masked_init_image.save("./masked_init.png")

pipe = StableDiffusionXLInpaintPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0",
    variant="fp16",
    torch_dtype=torch.float16,
    use_safetensors=True,
)
pipe.enable_attention_slicing()
pipe = pipe.to("cuda")

prompt = "a gray dog sitting on a park bench in front of trees"
image = pipe(
    prompt=prompt,
    image=masked_init_image,
    mask_image=inverted_mask
).images[0]

image.show()
image.save("./img.png")
```

We first use the mask to remove everything but the subject before feeding it into the inpaint pipeline. This makes the input image look like this: 
![masked_init (3)](https://github.com/huggingface/diffusers/assets/10781907/5d018f43-cf95-4a5b-9f7e-1d281fde6866)

The output image then looks like this:
![img (10)](https://github.com/huggingface/diffusers/assets/10781907/77589a31-cd6f-424f-bbe6-60fd1857364d)

It seems like the white background (which should have been masked out and not influence the image generation) ends up influencing the output. The model tries to incorporate this by making the image much whiter than you would otherwise expect.

Here's the result with SDXL inpainting v0.1. It's definitely better, but still looks like the white background is taking effect:
![inapint_model](https://github.com/huggingface/diffusers/assets/10781907/021c0103-79aa-41fd-8e86-4385ba028ae8)

Another way to deal with this is to set strength = 1.0, but that has its own drawbacks:
![img (9)](https://github.com/huggingface/diffusers/assets/10781907/e50c027a-b98f-4d31-ba8f-b65decbc8061)

Would it make sense that resizing the input + mask could cause the background to bleed through? If so is there a fix that would make sense for this context? I'd be happy to try and implement but could definitely use some direction 😅

**Comment by patrickvonplaten** [comment]
cc @williamberman here as well

**Comment by patrickvonplaten** [comment]
We should investigate this a bit more here. @yiyixuxu can you take this a look?

**Comment by babyta** [comment]
> when i use **content=original** in webui，so What corresponding changes should I make in the code of diffusers?

same question

**Comment by zengjie617789** [comment]
use sam-anything may help you a lot. The segment results make a different in results.

**Comment by crapthings** [comment]
same here, and why mask area touched?


![img](https://github.com/huggingface/diffusers/assets/1147704/eaa7f2a3-7105-4787-99d3-c65b175260fd)


![img_mask](https://github.com/huggingface/diffusers/assets/1147704/c22f0de0-1cb8-4220-ba5b-b7d083e74a46)


strength 0.95
<img width="642" alt="image" src="https://github.com/huggingface/diffusers/assets/1147704/8ed045d8-664c-4c0e-964b-d7e1e38822cf">

1.0
![image](https://github.com/huggingface/diffusers/assets/1147704/d280acdd-fd39-419f-80f5-4a62b4cd942d)


```
import torch
from diffusers.pipelines.stable_diffusion import safety_checker
from diffusers import StableDiffusionPipeline, StableDiffusionInpaintPipeline
from compel import Compel
from PIL import Image, ImageOps

from diffusers import DPMSolverMultistepScheduler

# override
def sc(self, clip_input, images): return images, [False for i in images]
safety_checker.StableDiffusionSafetyChecker.forward = sc

pipeline = StableDiffusionInpaintPipeline.from_single_file('./majicmixRealistic_v7-inpainting.safetensors', torch_dtype = torch.float16).to('cuda')
pipeline.scheduler = DPMSolverMultistepScheduler.from_config(pipeline.scheduler.config, algorithm_type = 'sde-dpmsolver++', use_karras_sigmas = True)
compel = Compel(tokenizer = pipeline.tokenizer, text_encoder = pipeline.text_encoder)

pipeline.enable_freeu(s1 = 0.9, s2 = 0.2, b1 = 1.2, b2 = 1.4)

prompt = 'photo++ of Luxury Loft: Sophisticated, upscale, modern interior, spacious layout, sleek design, floor-to-ceiling windows, minimalist decor, capturing a luxurious ambiance, crispy'
negative_prompt = 'ugly++, cartoon, animi, blurry++'

image = Image.open('./img.png')
mask = ImageOps.invert(Image.open('./img_mask.png'))

prompt_embeds = compel.build_conditioning_tensor(prompt)
negative_prompt_embeds = compel.build_conditioning_tensor(negative_prompt)

image = pipeline(
    image = image,
    mask_image = mask,
    prompt_embeds = prompt_embeds,
    negative_prompt_embeds = negative_prompt_embeds,
    width = 600,
    height = 960,
    num_inference_steps = 70,
    guidance_scale = 5,
    clip_skip = 2,
    strength = 1,
).images[0]

display(image)
```
